File: projects/06/Assembler/src/signed/SymbolTable.java
package signed;

import java.util.HashMap;
import java.util.Map;

/**
 * @author zhuangjy
 * @since 2021-3-26
 */
public class SymbolTable {
    private Map<String, Integer> map = new HashMap<>();

    public SymbolTable() {
        map.put("SP", 0);
        map.put("LCL", 1);
        map.put("ARG", 2);
        map.put("THIS", 3);
        map.put("THAT", 4);
        map.put("R0", 0);
        map.put("R1", 1);
        map.put("R2", 2);
        map.put("R3", 3);
        map.put("R4", 4);
        map.put("R5", 5);
        map.put("R6", 6);
        map.put("R7", 7);
        map.put("R8", 8);
        map.put("R9", 9);
        map.put("R10", 10);
        map.put("R11", 11);
        map.put("R12", 12);
        map.put("R13", 13);
        map.put("R14", 14);
        map.put("R15", 15);
        map.put("SCREEN", 16384);
        map.put("KBD", 24576);
    }

    public void addEntry(String symbol, int address) {
        map.put(symbol, address);
    }

    public boolean contains(String symbol) {
        return map.containsKey(symbol);
    }

    public int getAddress(String symbol) {
        return map.get(symbol);
    }
}

File: projects/06/Assembler-OC/Assembler-OC/SymbolTable.m
//
//  SymbolTable.m
//  Assembler-OC
//
//  Created by janezhuang on 2022/1/30.
//

#import "SymbolTable.h"

@interface SymbolTable ()
@property (nonatomic) NSMutableDictionary<NSString *, NSNumber *> *table;
@end


@implementation SymbolTable

- (instancetype)init {
    self = [super init];
    if (self) {
        self.table = @{
            @"SP" : @(0),
            @"LCL" : @(1),
            @"ARG" : @(2),
            @"THIS" : @(3),
            @"THAT" : @(4),
            @"R0" : @(0),
            @"R1" : @(1),
            @"R2" : @(2),
            @"R3" : @(3),
            @"R4" : @(4),
            @"R5" : @(5),
            @"R6" : @(6),
            @"R7" : @(7),
            @"R8" : @(8),
            @"R9" : @(9),
            @"R10" : @(10),
            @"R11" : @(11),
            @"R12" : @(12),
            @"R13" : @(13),
            @"R14" : @(14),
            @"R15" : @(15),
            @"SCREEN" : @(0x4000),
            @"KBD" : @(0x6000),
        }.mutableCopy;
    }
    return self;
}

- (void)addEntryWithSymbol:(NSString *)symbol address:(UInt32)address {
    self.table[symbol] = @(address);
}

- (BOOL)containsSymbol:(NSString *)symbol {
    NSNumber *number = [self.table objectForKey:symbol];
    return number != nil;
}

- (UInt32)getAddress:(NSString *)symbol {
    NSNumber *number = [self.table objectForKey:symbol];
    return [number unsignedIntValue];
}

@end
